src/Mvc/Mvc.ApiExplorer/src/EndpointMetadataApiDescriptionProvider.cs

```diff
@@ -0,0 +1,326 @@
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Internal;
+
+namespace Microsoft.AspNetCore.Mvc.ApiExplorer
+{
+    internal class EndpointMetadataApiDescriptionProvider : IApiDescriptionProvider
+    {
+        private readonly EndpointDataSource _endpointDataSource;
+        private readonly IServiceProviderIsService? _serviceProviderIsService;
+
+        // Executes before MVC's DefaultApiDescriptionProvider and GrpcHttpApiDescriptionProvider for no particular reason :D
+        public int Order => -1100;
+
+        public EndpointMetadataApiDescriptionProvider(EndpointDataSource endpointDataSource)
+            : this(endpointDataSource, null)
+        {
+        }
+
+        public EndpointMetadataApiDescriptionProvider(EndpointDataSource endpointDataSource, IServiceProviderIsService? serviceProviderIsService)
+        {
+            _endpointDataSource = endpointDataSource;
+            _serviceProviderIsService = serviceProviderIsService;
+        }
+
+        public void OnProvidersExecuting(ApiDescriptionProviderContext context)
+        {
+            foreach (var endpoint in _endpointDataSource.Endpoints)
+            {
+                if (endpoint is RouteEndpoint routeEndpoint &&
+                    routeEndpoint.Metadata.GetMetadata<MethodInfo>() is { } methodInfo &&
+                    routeEndpoint.Metadata.GetMetadata<IHttpMethodMetadata>() is { } httpMethodMetadata)
+                {
+                    // REVIEW: Should we add an ApiDescription for endpoints without IHttpMethodMetadata? Swagger doesn't handle
+                    // a null HttpMethod even though it's nullable on ApiDescription, so we'd need to define "default" HTTP methods.
+                    // In practice, the Delegate will be called for any HTTP method if there is no IHttpMethodMetadata.
+                    foreach (var httpMethod in httpMethodMetadata.HttpMethods)
+                    {
+                        context.Results.Add(CreateApiDescription(routeEndpoint, httpMethod, methodInfo));
+                    }
+                }
+            }
+        }
+
+        public void OnProvidersExecuted(ApiDescriptionProviderContext context)
+        {
+        }
+
+        private ApiDescription CreateApiDescription(RouteEndpoint routeEndpoint, string httpMethod, MethodInfo methodInfo)
+        {
+            // Swashbuckle uses the "controller" name to group endpoints together.
+            // For now, put all methods defined the same declaring type together.
+            string controllerName;
+
+            if (methodInfo.DeclaringType is not null && !IsCompilerGenerated(methodInfo.DeclaringType))
+            {
+                controllerName = methodInfo.DeclaringType.Name;
+            }
+            else
+            {
+                // If the declaring type is null or compiler-generated (e.g. lambdas),
+                // group the methods under a "Map" controller.
+                controllerName = "Map";
```

Comment:
A better fallback would be the application name. `IHostEnvironment.ApplicationName` or the assembly entry point name.